### ISSUE TITLE
DirectDraw works very poorly in Windows 11

### ISSUE THREAD

**Issue by Dwedit** [TASEmulators/fceux#541]
**Describe the bug**
DirectDraw is a graphics API that used to be fully functional in Windows XP and earlier, but ever since Desktop Composition (DWM) was added into Windows Vista, DirectDraw has been emulated by the operating system, with varying degrees of accuracy.

The quality of DirectDraw emulation has degraded over time.  It seems that in Windows 11, you *no longer get bilinear filtering when you stretch a rectangle*.  This is a basic feature of DirectDraw that has been in there since Windows 98, and Microsoft has now thrown it out for Windows 11.

So I would suggest moving from DirectDraw to something else.  I have experience working with Direct3D9, and that API is fully supported by both modern and ancient versions of Windows.

Is it okay if I start writing a D3D9 renderer for FCEUX?

**Comment by IPeluchito** [comment]
YES please!!! this emulator needs a lot more love in the non-Qt version, in any case if the developers don't agree you could create a fork...

**Comment by eugene-s-nesdev** [comment]
Another good reason to switch to D3D instead of DirectDraw (look at this issues please):

https://github.com/TASEmulators/fceux/issues/364
https://github.com/narzoul/DDrawCompat/issues/104

**Comment by Dwedit** [comment]
I've actually written something before that overrides DirectDraw and redirects its output to a D3D9 Texture DC.  But that was for a completely different motive (wanted to run cool pixel shaders on older DDraw games).

**Comment by TheBoxGame** [comment]
Czy ktoś wypuści update do FCEUX 2.6.5

**Comment by zeromus** [comment]
just smash the directdraw code and replace it with direct3d9. there's no need to keep directdraw around.

**Comment by Lincolnhsp** [comment]
> I've actually written something before that overrides DirectDraw and redirects its output to a D3D9 Texture DC. But that was for a completely different motive (wanted to run cool pixel shaders on older DDraw games). Also, it doesn't work on FCEUX in its current form.

Hello @Dwedit, about shaders you mentioned... just to contribute with some ideas... i´m using FCEUX with MagPie (which is also open source)... there are some interesting results combining both. 

Testing shaders over FCEUX is also very easy with Magpie (just editing a JSON with some parameters).

Maybe some idea brought with Magpie could be useful to implement the Direct3d render...

![image](https://user-images.githubusercontent.com/53958839/187096201-1f8e1504-99ee-4591-b325-229d47aaf0af.png)
GRT Geom - Interlaced
![image](https://user-images.githubusercontent.com/53958839/187096213-f0b46651-f36f-4098-b6aa-ad5c9c0b2354.png)

![image](https://user-images.githubusercontent.com/53958839/187096245-d09d1875-912e-483d-b9f6-a638ae308553.png)

CRT Hylian Shader
![image](https://user-images.githubusercontent.com/53958839/187096257-850ad0f2-ae8c-4437-95fe-6f2dbafda8fd.png)
![image](https://user-images.githubusercontent.com/53958839/187096310-baca72cf-f004-47c6-b58f-5b29c19b0f04.png)

Lanczos Adaptative Shader
![image](https://user-images.githubusercontent.com/53958839/187096363-c29ec92f-1bf3-4477-bb0d-47e33d4cb346.png)

XBRZ 6x Shader
![image](https://user-images.githubusercontent.com/53958839/187096384-2fbc81d2-5bc8-4cc2-87d2-a4a68772f603.png)

XBRZ Freescale Shader
![image](https://user-images.githubusercontent.com/53958839/187096407-fe483b4f-69f8-457b-b204-2dabb4ad8cff.png)

Anime4K Denoise + FSR
![image](https://user-images.githubusercontent.com/53958839/187096474-57caa82e-7c62-42b1-927c-4b607fc67781.png)

Best regards!

**Comment by zeromus** [comment]
Best idea from using magpie is to remove all support for upscaling from fceux entirely (at least the windows port) and expect people to use external tool to do it. It adds so much garbage code to an emulator to support all that junk internally

**Comment by eugene-s-nesdev** [comment]
You can try new DDrawCompat release here: https://github.com/narzoul/DDrawCompat/releases/tag/v0.4.0
This feature-request https://github.com/narzoul/DDrawCompat/issues/104 has been completed

**Comment by ClusterM** [comment]
Any progress? We really need this feature.

**Comment by warmCabin** [comment]
So glad I found this thread, I got a new computer recently and I was going *crazy!* I had better luck with [wined3d](https://fdossena.com/?p=wined3d/index.frag). It fixes the bilinear filtering issue without obliterating the default look and feel of Windows.
